specs/phase0/fork-choice.md

```diff
@@ -218,17 +267,30 @@ def filter_block_tree(store: Store, block_root: Root, blocks: Dict[Root, BeaconB
             return True
         return False
 
-    # If leaf block, check finalized/justified checkpoints as matching latest.
-    head_state = store.block_states[block_root]
+    current_epoch = compute_epoch_at_slot(get_current_slot(store))
+    voting_source = get_voting_source(store, block_root)
 
+    # The voting source should be at the same height as the store's justified checkpoint.
     correct_justified = (
         store.justified_checkpoint.epoch == GENESIS_EPOCH
-        or head_state.current_justified_checkpoint == store.justified_checkpoint
+        or voting_source.epoch == store.justified_checkpoint.epoch
```

Comment:
Shouldn't this be a check on checkpoint equivalence rather than than just epoch equivalence?

Once I see checkpoint X, I should only swap to another justified if it is *greater* than the epoch of X.

Reply:
If there are two conflicting checkpoints at the same height, we want LMD competition to take over. Otherwise, the order in which those conflicting checkpoints are received decides which fork we are on, which can cause deadlocks that require manual recovery.

Reply:
Lemma 4.11 in https://arxiv.org/pdf/2003.03052.pdf said that there will be no two conflicting justified checkpoints at the same epoch?

Reply:
Yes, that's correct. If we encounter the 1/3-slashable case of two conflicting justified checkpoints at the same height, we want the chain to keep progressing by resolving the tie using LMD competition.